core/commands/unixfs/ls.go

```diff
@@ -137,6 +137,10 @@ possible, please use 'ipfs ls' instead.
 			switch t {
 			case unixfspb.Data_File:
 				break
+			case unixfspb.Data_HAMTShard:
+				// We need a streaming ls API for this.
+				res.SetError(fmt.Errorf("cannot list large directories yet"), cmdkit.ErrNormal)
+				return
```

Comment:
I think it would be okay to have this without a streaming `ls`. It's just gonna take its time to run, just like dir-listings on the gateway for sharded directories.

The streaming `ls` is an optimization (an important one nevertheless).